src/Tools/Source/RunTests/TestHistoryManager.cs

```diff
@@ -0,0 +1,167 @@
+        //
+        // The phase name is used to filter the tests on the last passing build to only those that apply to the currently running phase.
+        //   Note here that 'phaseName' corresponds to the 'jobName' defined in our pipeline yaml file and the job name env var is not correct.
+        //   See https://developercommunity.visualstudio.com/t/systemjobname-seems-to-be-incorrectly-assigned-and/1209736
+        if (!TryGetEnvironmentVariable("SYSTEM_ACCESSTOKEN", out var accessToken)
+            || !TryGetEnvironmentVariable("SYSTEM_PHASENAME", out var phaseName))
+        {
+            Console.WriteLine("Missing required environment variables - skipping test history lookup");
+            return ImmutableDictionary<string, TimeSpan>.Empty;
+        }
+
+        // Use the target branch (in the case of PRs) or source branch to find the last successful build.
+        var targetBranch = Environment.GetEnvironmentVariable("SYSTEM_PULLREQUEST_TARGETBRANCH") ?? Environment.GetEnvironmentVariable("BUILD_SOURCEBRANCHNAME");
+        if (string.IsNullOrEmpty(targetBranch))
+        {
+            Console.WriteLine("Missing both PR target branch and build source branch environment variables - skipping test history lookup");
+            return ImmutableDictionary<string, TimeSpan>.Empty;
+        }
+
+        var credentials = new Microsoft.VisualStudio.Services.Common.VssBasicCredential(string.Empty, accessToken);
+
+        var connection = new VssConnection(s_projectUri, credentials);
+
+        using var buildClient = connection.GetClient<BuildHttpClient>();
+
+        Console.WriteLine($"Getting last successful build for branch {targetBranch}");
+        var adoBranch = $"refs/heads/{targetBranch}";
+        var builds = await buildClient.GetBuildsAsync2(project: "public", new int[] { RoslynCiBuildDefinitionId }, resultFilter: BuildResult.Succeeded, queryOrder: BuildQueryOrder.FinishTimeDescending, maxBuildsPerDefinition: 1, reasonFilter: BuildReason.IndividualCI, branchName: adoBranch, cancellationToken: cancellationToken);
+        var lastSuccessfulBuild = builds?.FirstOrDefault();
+        if (lastSuccessfulBuild == null)
+        {
+            // If this is a new branch we may not have any historical data for it.
+            ConsoleUtil.WriteLine($"Unable to get the last successful build for definition {RoslynCiBuildDefinitionId} and branch {targetBranch}");
+            return ImmutableDictionary<string, TimeSpan>.Empty;
+        }
+
+        using var testClient = connection.GetClient<TestResultsHttpClient>();
+
+        // API requires us to pass a time range to query runs for.  So just pass the times from the build.
+        var minTime = lastSuccessfulBuild.QueueTime!.Value;
+        var maxTime = lastSuccessfulBuild.FinishTime!.Value;
+        var runsInBuild = await testClient.QueryTestRunsAsync2("public", minTime, maxTime, buildIds: new int[] { lastSuccessfulBuild.Id }, cancellationToken: cancellationToken);
```

Comment:
FYI: my experience in runfo shows that getting test results is _extremely_ unreliable. We should not depend on this API not throwing 😦

Reply:
Fair - I will update this to catch exceptions.  I didn't do that originally since I wanted failures to fail the run while I worked on this.

Reply:
Yeah it's an attractive API but at scale it falls down. I have a number of retries in runfo around this API because not having them made all my analysis unreliable. It will also, frustratingly, fail then immediately pass for the exact same query.